### ISSUE TITLE
Feature Request: Scheduling Archival from the UI

### ISSUE THREAD

**Issue by BlipRanger** [ArchiveBox/ArchiveBox#578]
## Type 

 - [ ] General question or discussion
 - [x] Propose a brand new feature
 - [ ] Request modification of existing behavior or design

## What is the problem that your feature request solves

Currently scheduling ingestion of new urls requires writing a cron job external to the web UI (external to the docker container in my case) which isn't entirely ideal in a docker/self-contained setup. I believe this would be a nice convenience feature for users that might want to manage the entire operation of AB from within the web UI. 
 
## Describe the ideal specific solution you'd want, and whether it fits into any broader scope of changes

This feature would add a method for setting up scheduled pulls from various data sources via the web UI rather than only externally via cron. I specifically imagine at least a way to specify a RSS feed to be subscribed to that it can watch for new content from (something like Wallabag in my particular imagined use case). Technically I think this would involve a new menu/button in the UI and should dovetail with the internal scheduling processes already available. 

## How badly do you want this new feature?

 - [ ] It's an urgent deal-breaker, I can't live without it
 - [x] It's important to add it in the near-mid term future
 - [ ] It would be nice to have eventually

---

 - [x] I'm willing to contribute dev time / money to fix this issue
 - [x] I like ArchiveBox so far / would recommend it to a friend
 - [ ] I've had a lot of difficulty getting ArchiveBox set up

**Comment by pirate** [comment]
Yeah this is definitely on our mind, it probably won't be added for a couple versions but this is definitely something I've been planning.

It's blocked by adding a background queue system like Huey: https://github.com/ArchiveBox/ArchiveBox/issues/91

In the meantime I recommend using docker-compose instead of docker alone, as it allows you to declaratively define your scheduled imports all in one place (you can see the `docker-compose.yml` commented out section for an example of how to do that).

**Comment by BlipRanger** [comment]
Gotcha, I saw the future queuing system and that makes sense! And yes, currently using compose, so I'll look into doing that. Thanks!

**Comment by pirate** [comment]
Here's my proposed implementation of a new model to track scheduled imports:

```python3
class ScheduledTask(models.Models):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    
    name = models.CharField(max_length=128, required=True, unique=True)
    schedule = models.CharField(max_length=32, default='weekly')
    enabled = models.BooleanField(default=True)

    added = models.DateTimeField(auto_now_add=True)
    updated = models.DateTimeField(auto_now=True)

    runs = models.IntegerField(default=0, min_value=0, editable=False)
    last_output = models.CharField(max_length=1024, default='')
    last_start_ts = models.DateTimeField(default=None, null=True, blank=True)
    last_end_ts = models.DateTimeField(default=None, null=True, blank=True)
    
    class Meta:
        abstract = True


class ScheduledAdd(ScheduledTask):
    # main task parameters
    urls = models.TextField(max_length=4096, default='', db_index=True)
    tag = models.ManyToManyField(Tag)
    
    # add behavior flags
    depth = models.IntegerField(min_value=0, max_value=1, default=0)
    resnapshot = models.BooleanField(default=False)
    overwrite = models.BooleanField(default=False)
    index_only = models.BooleanField(default=False)
    update_all = models.BooleanField(default=False)

    extractors = models.CSVField(max_length=256, default='')
    parser = models.CharField(max_length=32, default='auto', choices=PARSER_CHOICES)
    
    @cached_property
    def source_filename(self) -> str:
        return f'{self.short_id}-scheduled-import.txt'

    def save(self, **kwargs):
        self.urls_str = self.urls_str.strip()

        assert self.urls_str or self.update_all, (
            'you must either pass some urls to import, or set the task to update'
            ' all existing URLS, otherwise it will do nothing')

        assert self.schedule in ('hour', 'day', 'week', 'month', 'year') or isValidCronSchedule(self.schedule)

        assert not (self.overwrite and self.resnapshot), (
            'When snapshotting a URL thats previously snapshotted, '
            'you may either overwrite it, or re-snapshot it, but not both')

        # some more validation here...
        save_text_as_source(self.urls, filename=self.source_filename)
        self.schedule()

    def schedule(self):
        method = 'system crontab' if USE_SYSTEM_CRON else 'archivebox scheduler'
        print(f'[*] Scheduling import {self.name} to run every {self.schedule} using {method}')

        # TODO: decide whether to support system cron at all, or enforce python scheduler
        if USE_SYSTEM_CRON:
            schedule(
                every=self.schedule,
                depth=self.depth,
                overwrite=self.overwrite,
                import_path=self.urls_source_file_path,
            )
        else:
            # TODO: update yacron/celery/huey/APScheduler etc. whatever scheduler we choose
            pass

    def run(self, force: bool=False):
        if (not self.enabled) and (not force):
            print(f'[!] Refusing to run scheduled import that is disabled: {self.name}')
            return None

        # TODO: enforce "at most once" or "at least once" concurrency somehow
        
        print(f'[+] [{timezone.now().isoformat()}] Running scheduled import {self.name}...\n')

        self.last_start_ts = timezone.now()
        self.runs += 1
        try:
            all_links, new_links = add(
                urls=Path(self.urls_source_file_path),
                tag=self.tag,
                depth=self.depth,
                update_all=self.update_all,
                index_only=self.index_only,
                overwrite=self.overwrite,
                extractors=self.extractors,
                parser=self.parser,
            )
            self.last_output = f'SUCCEEDED: {len(new_links)} new snapshots ({len(all_links)} total snapshots)'
        except BaseException as err:
            self.last_output = f'FAILED: {err.__class__.__name__} {err}'

        self.last_end_ts = timezone.now()
        self.save()

```

**Remaining TODOs:**

- [ ] figure out which python scheduler to use
  - huey (my current favorite)
  - celery (ugh...)
  - APScheduler (will require lots of manual models and concurrency control code)
  - yacron (not sure if it can be configured dynamically)
  - dramatiq (doesn't support sqlite)
- [ ] decide whether to continue supporting system crontab at all, or tear it out (imo we should just tear it out and move to using an internal scheduler)
- [ ] fork the scheduled task worker off the server process automatically on startup, so no need to run separate `archivebox schedule --foreground` process manually
- [ ] figure out how to enforce "at least once" or "at most once" concurrency model for scheduled tasks
